homeassistant/components/mqtt/client.py

```diff
@@ -573,12 +677,15 @@ async def _async_perform_subscriptions(
         def _process_client_subscriptions() -> list[tuple[int, int]]:
             """Initiate all subscriptions on the MQTT client and return the results."""
             subscribe_result_list = []
-            for topic, qos in subscriptions:
+            for topic, qos in subscriptions.items():
                 result, mid = self._mqttc.subscribe(topic, qos)
                 subscribe_result_list.append((result, mid))
                 _LOGGER.debug("Subscribing to %s, mid: %s, qos: %s", topic, mid, qos)
             return subscribe_result_list
 
+        subscriptions = copy.copy(self._pending_subscriptions)
+        self._pending_subscriptions.clear()
```

Comment:
```suggestion
        subscriptions = self._pending_subscriptions
        self._pending_subscriptions = {}
```

Do you need to copy or can you replace it with a fresh dict to avoid the copy?

Reply:
Should be safe since there is no await between and `self._pending_subscriptions` is only ever accessed in the main thread so it can't be touched by anything else until this function finishes and returns control to the event loop